File: backends/spring-boot/src/main/java/com/hellotimepro/springboot/service/UserService.java
package com.hellotimepro.springboot.service;

import com.hellotimepro.springboot.domain.UserEntity;
import com.hellotimepro.springboot.dto.Dtos.UpdateProfileRequest;
import com.hellotimepro.springboot.dto.Dtos.UserOut;
import com.hellotimepro.springboot.repository.UserRepository;
import com.hellotimepro.springboot.web.ApiException;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class UserService {
  private final UserRepository users;
  private final AvatarService avatars;
  private final MapperService mapper;

  public UserService(UserRepository users, AvatarService avatars, MapperService mapper) {
    this.users = users;
    this.avatars = avatars;
    this.mapper = mapper;
  }

  public UserOut toOut(UserEntity user) {
    return mapper.user(user);
  }

  @Transactional
  public UserOut updateProfile(UserEntity user, UpdateProfileRequest req) {
    if (req.nickname() == null && req.avatarId() == null) {
      throw ApiException.validation("至少提供一个字段", "body");
    }
    if (req.nickname() != null && !req.nickname().equals(user.getNickname())) {
      if (users.existsByNickname(req.nickname())) throw ApiException.conflict("昵称已被使用", "nickname");
      user.setNickname(req.nickname());
    }
    if (req.avatarId() != null) {
      if (!avatars.exists(req.avatarId())) throw ApiException.validation("头像 ID 不存在", "avatarId");
      user.setAvatarId(req.avatarId());
    }
    user.setUpdatedAt(OffsetDateTime.now(ZoneOffset.UTC));
    return mapper.user(users.save(user));
  }
}

File: fullstacks/rails/app/services/user_service.rb
module UserService
  module_function

  def to_dto(user)
    MapperService.user(user)
  end

  def update_profile(user, params)
    if params[:nickname].nil? && params[:avatarId].nil?
      raise ApiError.validation("至少提供一个字段", "body")
    end

    new_nickname = params[:nickname].nil? ? nil : Validation.nickname(params[:nickname])
    new_avatar = params[:avatarId].nil? ? nil : Validation.avatar_format(params[:avatarId])

    ActiveRecord::Base.transaction do
      if new_nickname && new_nickname != user.nickname
        raise ApiError.conflict("昵称已被使用", "nickname") if User.exists?(nickname: new_nickname)

        user.nickname = new_nickname
      end
      if new_avatar
        raise ApiError.validation("头像 ID 不存在", "avatarId") unless AvatarService.exists?(new_avatar)

        user.avatar_id = new_avatar
      end
      user.updated_at = Time.now.utc
      user.save!
    end
    MapperService.user(user)
  end
end
